Add Post on OTel Metrics Visualizations (#2841)
opensearch-project/project-website: _posts/2024-05-07-opentelemetry-metrics-visualization.md

---
layout: post
title:  "Using OpenSearch to visualize metrics ingested with OpenTelemetry"
authors:
 - kschnitter
date: 2024-07-16
categories:
 - technical-post
meta_keywords: 

> meta_description: 

Reviewer: @KarstenSchnitter - Please review and edit or add this meta to your blog: 

_meta_keywords:_ SAP Cloud Logging, Data Prepper ingestion, OpenTelemetry data models, OpenSearch Dashboards, observability plugin

_meta_description:_ Discover how SAP Cloud Logging created custom-built visualizations for analyzing metrics using the OpenTelemetry and Data Prepper ingestion tools in OpenSearch.

Author: This looks fine.

With our example setup, we can find data points similar to this (shortened) example:

```json
{
  "_source": {
    "name": "container.cpu.time",
    "description": "Total cumulative CPU time (sum of all cores) spent by the container/pod/node since its creation",
    "value": 120.353905,
    "flags": 0,
    "unit": "s",
    "startTime": "2024-04-10T06:49:56Z",
    "time": "2024-05-03T07:50:22.197391185Z",
    "kind": "SUM",
    "aggregationTemporality": "AGGREGATION_TEMPORALITY_CUMULATIVE",
    "isMonotonic": true,
    "serviceName": "otel-emailservice",
    "instrumentationScope.name": "otelcol/kubeletstatsreceiver",
    "instrumentationScope.version": "0.97.0",
    "resource.attributes.k8s@namespace@name": "example",
    "resource.attributes.k8s@deployment@name": "otel-emailservice",
    "resource.attributes.k8s@pod@name": "otel-emailservice-7795877686-r2lrt",
    "resource.attributes.k8s@container@name": "emailservice",
    "resource.attributes.service@name": "otel-emailservice"
  }
}
```

We can see that OTel separates metrics by the `name`.
Each point contains the generic `value` and `time`, which form the basic time series for the visualizations.

> This schema is a at first glance suboptimal for OpenSearch, where having a field with key `name` and value with the numerical value is easier accessible.

Reviewer 2: ```suggestion
This schema is a at first glance suboptimal for OpenSearch, where having a field with key `name` and a value containing the numerical value is more easily accessible.
```

---
layout: post

> title:  "The ABCs of semantic search in OpenSearch: Architectures, benchmarks, and combination strategies"

Author: This is missing. I will fix the title together with the date.

> title:  "The ABCs of semantic search in OpenSearch: Architectures, benchmarks, and combination strategies"

Reviewer 2: When we replace the title, please ensure that it's in sentence case (as it is currently) 😄

authors:
 - kschnitter
date: 2024-05-03
categories:
 - technical-post
meta_keywords: 
meta_description: 
excerpt: OpenSearch provides OpenTelemetry ingestion by using DataPrepper as ingestion tool. We explore, how to analyze and visualize metrics indexed by that route.
has_math: true
has_science_table: true
---

> OpenSearch is growing into a full observability platform supporting logs, metrics and traces.

Reviewer 3: Had exactly the same idea as the style-job

Using DataPrepper, it is capable of ingesting these three signal types via OpenTelemetry.
_Logs_ are supported well by the search index and other core features from OpenSearch and OpenSearch Dashboards.
_Traces_ have their own section within the Observability plugin providing flame graphs, service maps and much more.
The _metrics_ support in OpenSearch is geared towards the integration of an external Prometheus data source.
Analyzing metrics ingested via DataPrepper requires custom built visualizations.
This blog post is all about these visualizations and the underlying data model.

> It is built on the experience in SAP Cloud Logging service, where I created per-defined dashboards for the customers.

Reviewer 3: ```suggestion
It is built on the experience in [SAP Cloud Logging service](https://discovery-center.cloud.sap/serviceCatalog/cloud-logging), where I created per-defined dashboards for the customers.
```

We will do a less thorough approach and inspect example points to collect the necessary information.
With our example setup, we can find data points similar to this (shortened) example:

```json
{
  "_source": {
    "name": "container.cpu.time",
    "description": "Total cumulative CPU time (sum of all cores) spent by the container/pod/node since its creation",
    "value": 120.353905,
    "flags": 0,
    "unit": "s",
    "startTime": "2024-04-10T06:49:56Z",
    "time": "2024-05-03T07:50:22.197391185Z",
    "kind": "SUM",
    "aggregationTemporality": "AGGREGATION_TEMPORALITY_CUMULATIVE",
    "isMonotonic": true,
    "serviceName": "otel-emailservice",
    "instrumentationScope.name": "otelcol/kubeletstatsreceiver",
    "instrumentationScope.version": "0.97.0",
    "resource.attributes.k8s@namespace@name": "example",
    "resource.attributes.k8s@deployment@name": "otel-emailservice",
    "resource.attributes.k8s@pod@name": "otel-emailservice-7795877686-r2lrt",
    "resource.attributes.k8s@container@name": "emailservice",
    "resource.attributes.service@name": "otel-emailservice"
  }
}
```

We can see, that OpenTelemetry separates metrics by the `name`.

> Each point contains the generic `value` and `time`, which form the basic time series for our visualizations.

Reviewer 3: ```suggestion
Each point contains the generic `value` and `time`, which form the basic time series for the visualizations.
```
I'd rather use "the" instead of "our", but that's a matter of taste

We now switch back to the "Data" tab.

![TSVB Data Plain](/assets/media/blog-images/2024-05-07-opentelemetry-metrics-visualization/tsvb_data-max-value.png){:class="img-centered"}

Here we configure the data, we want to visualize.
We start simple and select aggregation "Max" of field "value".
This aligns well with the cumulative aggregation temporality.
In each interval the maximum (last) CPU time is placed as a dot in our diagram.
We group by a "Terms" aggregation on the K8s pod name to get the CPU times per pod.
To avoid over-crowding of our graph, we select Top 3 ordered by "Max of value" in descending direction.
This provides us with a time-line of CPU time by the top 3 pods.

If we were visualizing a metric of kind gauge, this would be it.
However, with our sum, there is more to be done.
The graph looks fine, but the arbitrary `startTime` differing between the pods makes the values hard to compare.
In the next step we calculate the rate of our sum.

![TSVB Data Positive Rate](/assets/media/blog-images/2024-05-07-opentelemetry-metrics-visualization/tsvb_data-positive-rate.png){:class="img-centered"}

We extend the aggregations by a derivative by unit 1s.

> This calculates the time share spent by the CPU on that pod.

Reviewer 2: In that case, would "the share of time spent" work?

We could have used the pod id as well, but pod names contain a unique suffix, so this is semantically identical.
Note, that the pod name is filled in by the kubeletstatsreceiver as resource attribute.

The count of containers can be added by clicking on the plus button with a very similar configuration.

![K8s Unique Container Count](/assets/media/blog-images/2024-05-07-opentelemetry-metrics-visualization/table_metric_unique-container-names.png){:class="img-centered"}

Note, that we provide custom labels to have nice table headings.
Any changes to our configuration are not applied automatically.
Use the "Update" button on the bottom right to sync the visualization.

We still need to provide separate lines per namespace.
This is achieved via a bucket aggregation.

![K8s Namespace Terms Aggregation](/assets/media/blog-images/2024-05-07-opentelemetry-metrics-visualization/table_metric_namespace-grouping.png){:class="img-centered"}

We choose "split rows" and a "Terms" aggregation for field `resource.attributes.k8s@namespace@name`.
We order our table by the number of pods.
We could also choose the number of containers or any other metrics we defined above.

> There is also an option for a complete separate metric.

Reviewer 3: ```suggestion
There is also an option for a completely separate metric.
```

We can also select the number of entries, we want to retrieve maximally.
The "Options" tab allows us to paginate this result by specifying the number of rows per page.
There we can also select a "total" function.
But in our case, only summing the number of pods would provide a correct number.
The container count might be double counting containers, that appear in different namespaces.
This is why this option was left unchecked.

Back in the "Data" tab, we could also opt to create a "missing values" bucket.
This can cost some query performance.
We can always check the query performance with the inspect dialog looking at "Requests" and the "Response" tab.

![Visualization Query Performance](/assets/media/blog-images/2024-05-07-opentelemetry-metrics-visualization/table_metric_performance.png){:class="img-centered"}

The value reported as `"took": 1787` indicates a latency of 1787ms.
We can always use this feature to analyze the impact of different aggregations and metrics configurations on the query runtime.

> This can be very helpful on complicated visualizations.

Reviewer 3: ```suggestion
This can be very helpful for complicated visualizations.
```

![Select Saved Search](/assets/media/blog-images/2024-05-07-opentelemetry-metrics-visualization/visualize_select-saved-search.png){:class="img-centered"}

We now need to provide several configurations.
Let us start with the x-axis, which needs to be a date histogram.
This divides the time interval selected on the upper right of the screen into time buckets.

![Date Histogram](/assets/media/blog-images/2024-05-07-opentelemetry-metrics-visualization/line_date-histogram.png){:class="img-centered"}

The default time field for OpenTelemetry metrics is `time`.
The minimum bucket length needs to be at least the reporting interval, to avoid empty buckets.
The actual length will be selected from the chosen interval.

We can now configure our y-axis.

![Calculating the CPU Time Slice](/assets/media/blog-images/2024-05-07-opentelemetry-metrics-visualization/line_y-axis.png){:class="img-centered"}

We want to use the derivative of each interval.
Since this is a parent aggregation, we need a time bucket aggregation as well.
Here we use the maximum, since the metric is a monotonically increasing counter.

In comparison with TSVB, we have no control over the derivation scale.

> That means, the derivative will be dependent on the interval length.

Reviewer 3: ```suggestion
That means the derivative will be dependent on the interval length.
```

This is something, we could not do in TSVB, which only supported grouping by one field.
We start with the pod name.

![Split by Pod Name](/assets/media/blog-images/2024-05-07-opentelemetry-metrics-visualization/line_split_pod-name.png){:class="img-centered"}

Note, that for ordering we are again restricted to the max aggregation.
We cannot sort by the derivative well, just like with TSVB.
After the pod, we split again by container name.

![Split by Container Name](/assets/media/blog-images/2024-05-07-opentelemetry-metrics-visualization/line_split_container-name.png){:class="img-centered"}

When you add the "split series" configuration, OpenSearch Dashboards will complain, that the Date Histogram needs to be the last visualization.
You can rearrange the bucket aggregations by drag and drop on the "=" icon.

With all that configuration done, we hit the Update button to inspect the graph.

![CPU Time by Container and Pod](/assets/media/blog-images/2024-05-07-opentelemetry-metrics-visualization/line_chart.png){:class="img-centered"}

We get a line of the cpu time used by time interval for each container.

> The labels consist of the pod name first and container name second.

Reviewer 3: ```suggestion
The labels consist of the pod name first and the container name second.
```

We can see the timestamp and cpu time used.
The labels for pod and container names are presented, but hard to recognize.
This is caused by the double aggregation.
If we used just pod names, this would be much better.
An alternative approach would be to use "split series" for container names and "split chart" for pod names.

Clicking on a data point brings up a filter creation dialog.

![Line Chart Point Filter](/assets/media/blog-images/2024-05-07-opentelemetry-metrics-visualization/line_data-point_filters.png){:class="img-centered"}

We can use these options to slice our data.
These filters are applied across the entire dashboard.

We have pushed our timeline visualization to the limits of the standard charts.
This gives us a good understanding of its capabilities.
It has its benefits for creating filters and grouping by multiple attributes.
But TSVB has easier options to format data.

### Outlook: Vega

We have explored, how to set up present OpenTelemetry metrics using TSVB and the standard visualizations.
Both approaches allowed the creation of useful charts providing powerful insights.
But we also encounter quite some limitations.

> To overcome those limits, OpenSearch Dashboards offers Vega as another powerful visualizations approach.

Reviewer 3: ```suggestion
To overcome those limits, OpenSearch Dashboards offers Vega as another powerful visualization approach.
```
